arangod/Aql/Functions.cpp

```diff
@@ -2466,6 +2466,27 @@ AqlValue Functions::CharLength(ExpressionContext* ctx, AstNode const&,
   return AqlValue(AqlValueHintUInt(length));
 }
 
+/// @brief function NORMALIZE
+AqlValue Functions::Normalize(ExpressionContext* ctx, AstNode const&,
+                              VPackFunctionParameters const& parameters) {
+  std::string utf8;
+  transaction::Methods* trx = &ctx->trx();
+  auto* vopts = &trx->vpackOptions();
+  AqlValue const& value = extractFunctionParameterValue(parameters, 0);
+
+  transaction::StringBufferLeaser buffer(trx);
+  arangodb::basics::VPackStringBufferAdapter adapter(buffer->stringBuffer());
+  size_t outlength;
+  char *dest;
+  
+  ::appendAsString(vopts, adapter, value);
+  
+  dest = TRI_normalize_utf8_to_NFC(buffer->c_str(), buffer->length(), &outlength);
+  auto ret = AqlValue(dest, outlength);
+  delete dest;
```

Comment:
There are multiple issues here:
* If the AqlValue in line 2485 cannot be constructed and throws an exception, the memory in `dest` is leaked.
* In line 2486, the memory for `dest` is released via `delete`, but it was allocated by `TRI_normalize_utf8_to_NFC` via `TRI_Allocate()`, which calls `malloc()`. So the memory must be freed using `TRI_Free()` here. Otherwise it is undefined behavior.
* `TRI_normalize_utf8_to_NFC` can return a nullptr in case it cannot allocate memory, which will then be passed to the constructor of AqlValue, which will result in undefined behavior, too.